# Assignment: Training and Evaluating a Small Transformer Language Model 

#### Course: Foundations of NLP

## 1. Setup

### 1.1 Clone the repository

In [1]:
!git clone https://github.com/karpathy/nanoGPT.git
%cd nanoGPT

f:\V.I.P\Uni\Foundations_NLP_assignment\nanoGPT


Cloning into 'nanoGPT'...


### 1.2 Install dependencies

In [2]:
!pip install torch numpy transformers datasets tiktoken wandb tqdm matplotlib

### 1.3 Prepare the dataset

In [3]:
%cd data/shakespeare_char
!python prepare.py
%cd ../..

f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\data\shakespeare_char
length of dataset in characters: 1,115,394
all the unique characters: 
 !$&',-.3:;?ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz
vocab size: 65
train has 1,003,854 tokens
val has 111,540 tokens
f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT


## 2. Baseline Training Run

In [4]:
import torch

if torch.cuda.is_available():
    device_type = "cuda"
else:
    device_type = "cpu"

print("Using device:", device_type)

Using device: cuda


In [ ]:
# Baseline configuration
out_dir = '../Experiments/out-shakespeare-baseline'
eval_interval = 250
eval_iters = 200
log_interval = 10

always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda' # change to 'cuda' if you have a GPU
compile = False # set True only on Linux with GPU

Run the baseline:

In [5]:
!python train.py ../config/train_shakespeare_char_baseline.py

^C


Overriding config with ../config/train_shakespeare_char_baseline.py:


Traceback (most recent call last):
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\train.py", line 77, in <module>
    exec(open('configurator.py').read()) # overrides from command line or config file
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 26, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '../config/train_shakespeare_char_baseline.py'


Overriding config with ../config/train_shakespeare_char_test.py:
# Baseline configuration
out_dir = '../out-shakespeare-baseline'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 2
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 1000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU
tokens per iteration will be: 16,384
found vocab_size = 65 (inside data\shakespeare_char\meta.pkl)
Initializing a new model from scratch
number of parameters: 3.57M
num decayed parameter tensors: 10, with 3,662,208 parameters
num non-decayed parameter tensors: 5, with 1,920 parameters
using fused AdamW: True
step 0: 

f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\train.py:196: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler(enabled=(dtype == 'float16'))
Traceback (most recent call last):
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\train.py", line 286, in <module>
    torch.save(checkpoint, os.path.join(out_dir, 'ckpt.pt'))
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\.venv\Lib\site-packages\torch\serialization.py", line 1003, in save
    with _open_zipfile_writer(f) as opened_zipfile:
         ~~~~~~~~~~~~~~~~~~~~^^^
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\.venv\Lib\site-packages\torch\serialization.py", line 865, in _open_zipfile_writer
    return container(name_or_buffer)  # ty

Overriding config with ../config/train_shakespeare_char_test.py:
# Baseline configuration
out_dir = '../Experiments/out-shakespeare-test'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 2
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 1000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU
tokens per iteration will be: 16,384
found vocab_size = 65 (inside data\shakespeare_char\meta.pkl)
Initializing a new model from scratch
number of parameters: 3.57M
num decayed parameter tensors: 10, with 3,662,208 parameters
num non-decayed parameter tensors: 5, with 1,920 parameters
using fused AdamW: True


f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\train.py:196: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler(enabled=(dtype == 'float16'))
Traceback (most recent call last):
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\train.py", line 286, in <module>
    torch.save(checkpoint, os.path.join(out_dir, 'ckpt.pt'))
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\.venv\Lib\site-packages\torch\serialization.py", line 1003, in save
    with _open_zipfile_writer(f) as opened_zipfile:
         ~~~~~~~~~~~~~~~~~~~~^^^
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\.venv\Lib\site-packages\torch\serialization.py", line 865, in _open_zipfile_writer
    return container(name_or_buffer)  # ty

### Generate text from the trained model

In [6]:
!python sample.py --out_dir=../Experiments/out-shakespeare-baseline --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-baseline
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...


ANGELO:
And cowards it be straighted but of evils,
So many gates shall this us arm to be determined
Twice with false of our peaces: make it me of mine,
Your best remedy accessaries, drove so far
Will you go with self-increase.

ISABELLA:
I would have pleased with me no more sorrow
To this odd to age on your what evils we are done,
Even shame strong of his guard and thrust for treason!

DUKE VINCENTIO:
If he bring your horse-be properobation and be
To Elizabething straight no more stones of the 
---------------

Menenius, I must like a whisperator, I will not not be
well thrown to the place. You will not with your son: show you your
names for a care-like ever when moved the part
before your virtues: so the hour anger, as you come,
as you never no doubt need o

## 3. Hyperparameter Experiments

Design  and  run  at  least  four  additional  experiments,  varying  one  hyperparameter  at  a  time relative to the baseline. You must cover at least three of the following categories:

| Category | Hyperparameter | Suggested values to try |
| - | - | - |
| Learning rate | learning_rate | 1e-4, 5e-4, 1e-3 (baseline), 5e-3, 1e-2 |
| Model depth | n_layer |2, 4, 6 (baseline), 8 |
| Model width | n_embd (n_head) | 128 (4 heads), 256 (4 heads), 384 (6 heads, baseline) |
| Context length | block_size | 64, 128, 256 (baseline) |
| Regularisation | dropout | 0.0, 0.1, 0.2 (baseline), 0.4 |
| Training duration | max_iters | 1000, 2500, 5000 (baseline), 10000 |

For each experiment, create a separate config file (e.g., config/train_shakespeare_char_lr5e4.py)  and  a  separate  out_dir  so  that  results  are  not overwritten.

Important: Keep all other hyperparameters fixed when varying one. Record the exact
configuration for every run.

In [ ]:
# Low Learningrate configuration
out_dir = '../Experiments/out-shakespeare-low_lr'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-4 #CHANGE FROM BASELINE
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_low_lr.py

In [ ]:
# 2 Layers configuration
out_dir = '../Experiments/out-shakespeare-2_layers'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 2 #CHANGE FROM BASELINE
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_2_layers.py

In [ ]:
# 4 Layers configuration
out_dir = '../Experiments/out-shakespeare-4_layers'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 4 #CHANGE FROM BASELINE
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_4_layers.py

In [ ]:
# 64 Context configuration
out_dir = '../Experiments/out-shakespeare-64_context'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 64 #CHANGE FROM BASELINE

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_64_context.py

python: can't open file 'f:\\V.I.P\\Uni\\Foundations_NLP_assignment\\train.py': [Errno 2] No such file or directory


In [ ]:
# 128 Context configuration
out_dir = '../Experiments/out-shakespeare-128_context'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 128 #CHANGE FROM BASELINE

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_128_context.py

In [ ]:
# 128 width configuration
out_dir = '../Experiments/out-shakespeare-128_width'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 4 #CHANGE FROM BASELINE
n_embd = 128 #CHANGE FROM BASELINE
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_128_width.py

In [ ]:
# 256 width configuration
out_dir = '../Experiments/out-shakespeare-256_width'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 4 #CHANGE FROM BASELINE
n_embd = 256 #CHANGE FROM BASELINE
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_256_width.py

In [ ]:
# 1000 Iterations configuration
out_dir = '../Experiments/out-shakespeare-1000_iters'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 1000 #CHANGE FROM BASELINE
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_1000_iters.py

In [ ]:
# 2500 Iterations configuration
out_dir = '../Experiments/out-shakespeare-2500_iters'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 2500 #CHANGE FROM BASELINE
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py ../config/train_shakespeare_char_2500_iters.py

## 4. Evaluation and Reporting

4.1 Quantitative evaluation

For each run, record:

• Final training loss

• Final validation loss

• Training time (wall-clock)

In [9]:
def readOutLogFile(folderPath, fileName):
    import os
    import numpy as np
    from string import printable

    logFileContent = ""

    filePath = os.path.join(folderPath, fileName)

    with open(filePath, 'r') as logFile:

        logFileContent = logFile.read()

    logData = {}
    logData["path"] = filePath
    logData["Experiment"] = fileName.replace("_", " ").replace(".log", "")
    logData["Iters"] = 5000
    
    itterationTimes = 0
    evaluationTimes = 0
    eval_interval = 250
    lastIteration = 0

    lines = logFileContent.split("\n")
    for line in lines:
        lineContext = ''.join(char for char in line if char in printable)
        lineContext = lineContext.replace(" ", "").replace("#CHANGEFROMBASELINE", "")
        if len(lineContext) > 1:
            if lineContext.startswith("learning_rate"):
                logData["LR"] = f"{float(lineContext.replace("learning_rate=", "")):.1e}"
            if lineContext.startswith("n_layer"):
                logData["Layers"] = int(lineContext.replace("n_layer=", ""))
            if lineContext.startswith("n_embd"):
                logData["Embd"] = int(lineContext.replace("n_embd=", ""))
            if lineContext.startswith("n_head"):
                logData["n_head"] = int(lineContext.replace("n_head=", ""))
            if lineContext.startswith("block_size"):
                logData["Block"] = int(lineContext.replace("block_size=", ""))
            if lineContext.startswith("dropout"):
                logData["Dropout"] = float(lineContext.replace("dropout=", ""))
            if lineContext.startswith("max_iters"):
                logData["Iters"] = int(lineContext.replace("max_iters=", ""))
            if lineContext.startswith("eval_interval"):
                eval_interval = int(lineContext.replace("eval_interval=", ""))
            if lineContext.startswith("step"):

                stepData = {}
                
                lineContextParts = lineContext.replace(":", ",").split(",")

                stepData["step"] = int(lineContextParts[0].replace("step", ""))
                stepData["trainloss"] = float(lineContextParts[1].replace("trainloss", ""))
                stepData["valloss"] = float(lineContextParts[2].replace("valloss", ""))

                if not ("steps" in logData):
                    logData["steps"] = []
                
                logData["steps"].append(stepData)
            if lineContext.startswith("iter"):

                iterData = {}
                
                lineContextParts = lineContext.replace(":", ",").split(",")

                iterData["iter"] = int(lineContextParts[0].replace("iter", ""))
                iterData["loss"] = float(lineContextParts[1].replace("loss", ""))
                iterData["time"] = float(lineContextParts[2].replace("time", "").replace("ms", ""))
                iterData["mfu"] = float(lineContextParts[3].replace("mfu", "").replace("%", ""))

                if not ("iter" in logData):
                    logData["iter"] = []

                if not (iterData["iter"] % eval_interval == 0):
                    itterationTimes += iterData["time"] * (iterData["iter"] - lastIteration)
                else:
                    evaluationTimes += iterData["time"]
                    if len(logData["iter"]) > 0 and iterData["iter"] - lastIteration - 1 >= 1:
                        itterationTimes += logData["iter"][-1]["time"] * (iterData["iter"] - lastIteration - 1)
                    
                logData["iter"].append(iterData)
                lastIteration = iterData["iter"]

    logData["Time (min)"] = round((itterationTimes + evaluationTimes) / 1000 / 60, 2)
    logData["Train Loss"] = logData["steps"][-1]["trainloss"]
    logData["Val Loss"] = logData["steps"][-1]["valloss"]

    return logData

In [11]:
logsFolderPath = "../logs/"
logFileNames = ["baseline.log", "low_lr.log", "2_layers.log", "4_layers.log", "64_context.log", "128_context.log", "128_width.log", "256_width.log", "2500_iters.log", "1000_iters.log"]
tableColumns = ["Experiment", "LR", "Layers", "Embd", "Dropout", "Iters", "Train Loss", "Val Loss", "Time (min)"]

headderStr = "|"
seperatorStr = "|"
for column in tableColumns:
    headderStr += " " + "{:<11}".format(column) + " |"
    seperatorStr += "-" * 13 + "|"

print(headderStr)
print(seperatorStr)

for logFileName in logFileNames:
    logFileData = readOutLogFile(logsFolderPath, logFileName)
    
    rowStr = "|"

    for column in tableColumns:
        if column in logFileData:
            rowStr += " " + "{:<11}".format(str(logFileData[column])) + " |"
        else:
            rowStr += " |"

    print(rowStr)

| Experiment  | LR          | Layers      | Embd        | Dropout     | Iters       | Train Loss  | Val Loss    | Time (min)  |
|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| baseline    | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.6232      | 1.6978      | 44.65       |
| low lr      | 1.0e-04     | 6           | 384         | 0.2         | 5000        | 1.1558      | 1.5021      | 44.63       |
| 2 layers    | 1.0e-03     | 2           | 384         | 0.2         | 5000        | 1.1023      | 1.4854      | 16.22       |
| 4 layers    | 1.0e-03     | 4           | 384         | 0.2         | 5000        | 0.8542      | 1.5319      | 29.97       |
| 64 context  | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 1.1576      | 1.4823      | 11.82       |
| 128 context | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.9449      | 1.50

#### Our Results:

| Experiment  | LR          | Layers      | Embd        | Dropout     | Iters       | Train Loss  | Val Loss    | Time (min)  |
|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| baseline    | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.6232      | 1.6978      | 44.65       |
| low lr      | 1.0e-04     | 6           | 384         | 0.2         | 5000        | 1.1558      | 1.5021      | 44.63       |
| 2 layers    | 1.0e-03     | 2           | 384         | 0.2         | 5000        | 1.1023      | 1.4854      | 16.22       |
| 4 layers    | 1.0e-03     | 4           | 384         | 0.2         | 5000        | 0.8542      | 1.5319      | 29.97       |
| 64 context  | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 1.1576      | 1.4823      | 11.82       |
| 128 context | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.9449      | 1.5016      | 23.23       |
| 128 width   | 1.0e-03     | 6           | 128         | 0.2         | 5000        | 1.2719      | 1.5157      | 12.45       |
| 256 width   | 1.0e-03     | 6           | 256         | 0.2         | 5000        | 0.9703      | 1.4646      | 24.06       |
| 2500 iters  | 1.0e-03     | 6           | 384         | 0.2         | 2500        | 0.9565      | 1.4929      | 22.86       |
| 1000 iters  | 1.0e-03     | 6           | 384         | 0.2         | 1000        | 1.2706      | 1.5231      | 9.61        |

### 4.2 Loss curves

Create a **Python script** that:

• Parses the training logs for each run (nanoGPT prints step, train loss, val loss to stdout).

• Plots **training loss vs. step** for all runs in a single figure (one line per run, with a legend).

• Plots **validation loss vs. step** in a second figure.

Save these as .png or .pdf files. **Tip:** Redirect output to a log file during training:

In [ ]:
# OUR PYTHON CODE

### 4.3 Qualitative evaluation

For **each** experiment, generate at least 3 text samples (500 tokens each) and include them in an
appendix. In the main body of your report, discuss:

• Which configurations produce the most coherent or "Shakespeare-like" text?

• At what point do you see clear signs of underfitting (e.g., gibberish output)?

• Do you observe any signs of overfitting (low train loss, high val loss, repetitive output)?

• Is there a noticeable difference between models with different context lengths?

In [12]:
!python sample.py --out_dir=../Experiments/out-shakespeare-baseline --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-baseline
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...


ANGELO:
And cowards it be straighted but of evils,
So many gates shall this us arm to be determined
Twice with false of our peaces: make it me of mine,
Your best remedy accessaries, drove so far
Will you go with self-increase.

ISABELLA:
I would have pleased with me no more sorrow
To this odd to age on your what evils we are done,
Even shame strong of his guard and thrust for treason!

DUKE VINCENTIO:
If he bring your horse-be properobation and be
To Elizabething straight no more stones of the 
---------------

Menenius, I must like a whisperator, I will not not be
well thrown to the place. You will not with your son: show you your
names for a care-like ever when moved the part
before your virtues: so the hour anger, as you come,
as you never no doubt need o

In [13]:
!python sample.py --out_dir=../Experiments/out-shakespeare-low_lr --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-low_lr
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...


Nurse:
Ready, will you be better you but our hands.
3 KING HENRY VI

KING RICHARD II:
Why dost was, away, my lord, and God's king,
Let friar Bianca will be cause.

RICHARD:
O players, or that Bolingbroke!

NORTHUMBERLAND:
Thou liest my heart'st that speak you love's king.

YORK:
What thy world that more would do prize thy blood,
Which now means so more for the king thrust for the
rise my mother.

Post:
Your Prince, Sir, and be a crown discoved a curse
That thou best a cravel in his charged vow,
---------------

Men pardon me such of mine and master and see
On the prince of the stroke of head and us.

ISABELLA:
These comes that lords and so that would shall be be
thou to make the excolaim of my sovereign
Even to me from her better of chance:
And let he needless

In [14]:
!python sample.py --out_dir=../Experiments/out-shakespeare-2_layers --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-2_layers
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 3.57M
Loading meta from data\shakespeare_char\meta.pkl...


ANGELO:
And cowards it be stranged, and you well.
Once the gates shall this unquire our complainted
Away with himself a promise
And make it in the sweet liberty is enemy.
We ladiest drop the deed known, on the sea and no more.
How now, what stay have you tyrants me well;
It is all the midwife of all of your whole I beseech
A dogress of his earth-owledge; the dear mother,
If I am not some more pince of my sovereign,
So much I see as in the wards of his grave?

GLOUCESTER:
And I, my lord, I can y
---------------

Men pass'd, I shall be so.

KING RICHARD II:
See, my lord, I may not; and be her than thy father's
cousin's most that lord
this son may be so seen as when my statue passage, I say, in
the seasons of the creatures of blood, as I know thou liest,
now, fo

In [15]:
!python sample.py --out_dir=../Experiments/out-shakespeare-4_layers --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-4_layers
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 7.11M
Loading meta from data\shakespeare_char\meta.pkl...


KING RICHARD II:
Say that stay a sea boy of that hath slept,
Takes a thousand queen. Come on, come, come, my father,
I would not stand for the court. Come, cousin, let's go.

STANLEY:
Why, what I know.

GLOUCESTER:
I am at all the malice of my grieving lord?

KING HENRY VI:
Alas, whom I should that move have her eyes so here bound,
The shame strong of her burthern and the field;
And therefore you would find me for your hands,
Which, you say advances him for his blood and tears,
It may not call 
---------------

Men pardons me yourself that you should show.

LEONTES:
Not that, my lord;
I have no witness, with confessions to case,
I must speak apace stronger.

ANTIGONUS:
I must be to desire her; but what's the reason was
In the morning.

LEONTES:
What's the oxl

In [16]:
!python sample.py --out_dir=../Experiments/out-shakespeare-64_context --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-64_context
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...


Clown:
Read your father's blood: yet but one way
But to meet the sun and her heart to be determine,
And so hateful a wizon made what to thy heart
Will discover mine enemy, will it think you,
He will not pity and sell in your holdship
Be much but back'd, why should she red at law,
Though I shall be leg to such death by the time
And can not record of my heart
To know thrust for the great shadow your poor field
And grant unresh upon your dischards.

POLIXENES:
This courtesy?

SIR STEPHEN SCROOP:
O
---------------

Men pardon, marry, sir, what you say?

ESCALUS:
Some honour throat me stand in this law;
If he would have made the mighty motions,
That which shall be beheld the duke to be so
As known their soldiers are fearful sours,
Or, nothing is gone of her eye

In [17]:
!python sample.py --out_dir=../Experiments/out-shakespeare-128_context --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-128_context
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...


ANGELO:
And come, my lord,
And after you, my lord, that we have
My arm that use your tragedies was, and the confound
An only paints much in his that would be embraced?
And will I make over-pain with him to the sea like down,
And she could not speak i' the years of Henry
To this loss pleasure mother'd and so did Corioli Stanley?
What cousin Bushy? O my house
To know thrust for triumph! muffled you not the people?

MARIANA:
This is strong and adopted.

ESCALUS:
Go to know and give thee from the w
---------------

Men pardon, marry, I will give me before him.

DUKE OF YORK:
Art thou stay, in God's wife and Tybalt?

YORK:
Why, come, my lords, that is the sea of me to my state?
Canst thou so know? You must not hear me Rome,
To the gold give thee her like a nee

In [18]:
!python sample.py --out_dir=../Experiments/out-shakespeare-128_width --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-128_width
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 1.19M
Loading meta from data\shakespeare_char\meta.pkl...


ANGELO:
And come, which you that you bear to take him my dage?

MENENIUS:
No wise, therefore, how ever your head,
The more man of the officer both of my lord.
Which, since will it the overtain.

POMPEY:
Well, by selfsance, and the mild with a cheeks
As they have not not do this death:
I will met ake out--so which I beseech your right,
Than you passage his business noble to such;
And there's fearful of a fair order, for the good
Hasting in a point this great hoir courter's name?

PETER:
Why shal
---------------

Men pardon me such shall have hand have myself,
Something that take the crown of thy subjects.
I must friend should sound my cast thy peace
Of his pardon's love, induction of this peeding.

DUKE OF AUMERLE:
Alas, not come, sir, sir, thou love to entre

In [19]:
!python sample.py --out_dir=../Experiments/out-shakespeare-256_width --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-256_width
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 4.74M
Loading meta from data\shakespeare_char\meta.pkl...


DUCHESS OF YORK:
She shall see his heart of evil.
O day with hands artic thou quickent?
We that are away with himself thought
Your proof in his that would be embired:
You hear them in the seat.

WARWICK:
And all well and tears them with me with their life
If not stumbled peevish with thy world that
To Warwick do evil the king, this day had not power
And struck them to the first; would he there is not
The season of the coron-bedage, be discharged
With a help hois courtesy tears,--the heart care,
---------------

Men pardon, marry, I will be thy son,
Be his mercy, by weak that hath stain'd his life.

KING RICHARD III:
It shall so, madam, so it is in stronger.

ABHORSON:
Let me see the shink of the seas, not he lives,
And see him that sceptress of love to dear 

In [20]:
!python sample.py --out_dir=../Experiments/out-shakespeare-1000_iters --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-1000_iters
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...

And they be desired, is not the disposities of all
On the colderers: they have queen by arms last,
But when I defend the king, and the office
Become the world embirests, and that Hereford the head
Which is better well in the little trouble
Nor special the tyrant to see Bolingbroke,
And the self that more not our Corioli with this
That I have done of my heart
To know thrust first; and there's tears of a word
Marriage home. Her such at your army speeches
Where hoice must are no much soldiers care,
---------------

MenEnius, be a younger.
Why death! have you how? what the men reason; and
You have clough one with confessests the might. And you have been
Master along me overeign pass'd of this poverty.

VOLUMNIA:
When is that not come?

MENENIUS:
No, if I condem

In [21]:
!python sample.py --out_dir=../Experiments/out-shakespeare-2500_iters --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = ../Experiments/out-shakespeare-2500_iters
Overriding: device = cuda
Overriding: num_samples = 3
Overriding: max_new_tokens = 500
number of parameters: 10.65M
Loading meta from data\shakespeare_char\meta.pkl...


Clown:
So, who is her lady? that show our honours?

AUTOLYCUS:

MENENIUS:
No more, my lord: let me say your honour, with the
paor to the patrician! What!

AUTOLYCUS:
Evermore a very drev ten to be longer
in your pelicious ear.

AUTOLYCUS:
Now are near he, you lord.

Clown:
Yet all the matter of the moon will be Corioli with
the right o' the contracted his business; three affection
is not to thy fledge, to fight upon of.

Clown:
Hark, by any admiration!

Clown:
He stands and offends off and vill
---------------

Men pardon me, you shall have hanged to myself.

Lord:
Well, madam, what same I will such as this,
So fasty that these brother should be joy.

Lord:
Tell me your graces!

KING EDWARD IV:
To who keep thou against thy body forget
That thou liest in Fr

### 4.4 Analysis questions

Answer the following in your write-up (2–4 sentences each):

• **1.**  What  happens  when  the  learning  rate  is  too  high?  Too  low?  How  does  this  manifest  in  the
loss curve and the generated text?

• **2.**  What  is  the  relationship  between  model  size  (layers/embedding)  and  validation  loss?  Is
bigger always better at this scale and dataset size?

• **3.** What role does dropout play? Compare the train/val loss gap across dropout values.

• **4.** Based on your experiments, what configuration would you recommend for this dataset, and
why?

#### 5. Bonus (optional)

Choose one or both of the following:

• **A.** Train on a **custom dataset** of your choice (minimum 500KB of text). Describe the dataset,
any preprocessing you did, and compare sample quality with Shakespeare.

• **B.**  Replace  the  character-level  tokeniser  with  your  own  **Byte  Pair  Encoding  (BPE)**  tokeniser.
Train it on the Shakespeare corpus (or your custom dataset), integrate it into the nanoGPT data
pipeline,  and  compare  training  dynamics  and  generation  quality  against  the  character-level
baseline.